# **NAIVE Bayes algorithm**

**Formula:**

$$P(A|B) = \frac{P(B|A)P(A)}{P(B)}$$

where A and B are events and P(B) != 0
- `P(A|B)` is the probability of event A occurring given that event B has already occurred.
- `P(B|A)` is the probability of event B occurring given that event A has already occurred.
- `P(A)` and P(B) are the probabilities of events A and B occurring independently.
- `P(B)` is the probability of event B occurring.
- `P(A|B)` is the posterior probability.
- `P(B|A)` is the likelihood.

### **Naive Bayes Algorithm:**

Naive Bayes Algorithm is based on Bayes Theorem. It is defined as:

$$P(y|x_1,x_2,...,x_n) = \frac{P(x_1,x_2,...,x_n|y)P(y)}{P(x_1,x_2,...,x_n)}$$

where y is the class variable and x1, x2, ..., xn are the features.

The algorithm assumes that the features are independent of each other. So, the above equation can be written as:

$$P(y|x_1,x_2,...,x_n) = \frac{P(x_1|y)P(x_2|y)...P(x_n|y)P(y)}{P(x_1,x_2,...,x_n)}$$

The denominator is constant for a given input. So, the equation can be written as:

$$P(y|x_1,x_2,...,x_n) \propto P(x_1|y)P(x_2|y)...P(x_n|y)P(y)$$

The class with the highest probability is the output of the algorithm.

### **1. Import Libraries:**

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.naive_bayes import GaussianNB, MultinomialNB, BernoulliNB
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.datasets import load_iris

### **2. Load Data:**

In [2]:
# load the dataset
iris = load_iris()

### **3. Separate Features (X) and Target(y):**

In [3]:
X = iris.data
y = iris.target

### **4. Split Data into Training and Testing Sets:**

In [4]:
# train test split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

### **5. Build & Train the Model:**

#### **`Gaussion Naive Bayes`:**

In [5]:
# model initialize
gnb = GaussianNB()

# train the model
gnb.fit(X_train, y_train)

GaussianNB()

### **6. Make Predictions:**

In [6]:
# predict the test data
y_pred = gnb.predict(X_test)

### **7. Evaluate the `Gaussian Naive Bayes` Model:**

In [7]:
# evaluate the model
print("Accuracy Score: ", accuracy_score(y_test, y_pred))
print("Confusion Matrix: \n", confusion_matrix(y_test, y_pred))
print("Classification Report: \n", classification_report(y_test, y_pred))

Accuracy Score:  0.9777777777777777
Confusion Matrix: 
 [[19  0  0]
 [ 0 12  1]
 [ 0  0 13]]
Classification Report: 
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        19
           1       1.00      0.92      0.96        13
           2       0.93      1.00      0.96        13

    accuracy                           0.98        45
   macro avg       0.98      0.97      0.97        45
weighted avg       0.98      0.98      0.98        45



The above code is using the `Gaussian Naive Bayes algorithm` to classify the Iris dataset. Here's a breakdown of the code:

- The Iris dataset is loaded, and the features (`X`) and target variable (`y`) are separated.
- The data is split into training and test sets using `train_test_split`.
- The Gaussian Naive Bayes model is initialized and trained on the training data.
- The trained model is used to make predictions on the test data.
- The model's performance is evaluated by calculating the accuracy score, confusion matrix, and classification report.

**Observations from the Output:**
| Metric | Class 0 | Class 1 | Class 2 | Macro Avg | Weighted Avg | Remarks |
|--------|---------|---------|---------|-----------|--------------|---------|
| Precision | 1.00 | 1.00 | `0.93` | 0.98 | 0.98 | All instances predicted as class 0 and 1 were correct. For class 2, 93% of instances predicted were correct. |
| Recall | 1.00 | `0.92` | 1.00 | 0.97 | 0.98 | All actual instances of class 0 and 2 were identified. For class 1, 92% of actual instances were identified. |
| F1-Score | 1.00 | 0.96 | 0.96 | 0.97 | 0.98 | High F1-score across all classes indicates a good balance between precision and recall. |
| Support | 19 | 13 | 13 | --- | 45 | Balanced support across classes indicates a balanced test set. |
| Accuracy | --- | --- | --- | --- | 0.98 | High overall accuracy indicates that the model correctly classified 98% of all instances. |
| Confusion Matrix | 19/19 | `12`/13 | 13/13 | --- | --- | All instances of class 0 were correctly classified. For class 1, one instance was misclassified as class 2. All instances of class 2 were correctly classified. |

**In summary**, the Gaussian Naive Bayes model performed very well on the Iris dataset, with high accuracy and precision for all classes. The model had a slight difficulty with class 1 recall and class 2 precision, but overall, the performance was excellent. The model's high F1-score indicates a good balance between precision and recall. The support shows that the classes are balanced in the test set. The high overall accuracy of 0.98 indicates that the model correctly classified 98% of all instances. The confusion matrix shows that the model had a slight difficulty in distinguishing between class 1 and class 2, leading to one instance of class 1 being misclassified as class 2.

### **`Multinomial Naive Bayes`:**

In [8]:
# model initialize
mnb = MultinomialNB()

# train the model
mnb.fit(X_train, y_train)

# predict the test data
y_pred = mnb.predict(X_test)

# evaluate the model
print("Accuracy Score: ", accuracy_score(y_test, y_pred))
print("Confusion Matrix: \n", confusion_matrix(y_test, y_pred))
print("Classification Report: \n", classification_report(y_test, y_pred))

Accuracy Score:  0.9555555555555556
Confusion Matrix: 
 [[19  0  0]
 [ 0 12  1]
 [ 0  1 12]]
Classification Report: 
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        19
           1       0.92      0.92      0.92        13
           2       0.92      0.92      0.92        13

    accuracy                           0.96        45
   macro avg       0.95      0.95      0.95        45
weighted avg       0.96      0.96      0.96        45



**Observations from the Output:**
| Metric | Class 0 | Class 1 | Class 2 | Macro Avg | Weighted Avg | Remarks |
|--------|---------|---------|---------|-----------|--------------|---------|
| Precision | 1.00 | 0.92 | `0.92` | 0.95 | 0.96 | All instances predicted as class 0 were correct. For classes 1 and 2, 92% of instances predicted were correct. |
| Recall | 1.00 | 0.92 | `0.92` | 0.95 | 0.96 | All actual instances of class 0 were identified. For classes 1 and 2, 92% of actual instances were identified. |
| F1-Score | 1.00 | 0.92 | 0.92 | 0.95 | 0.96 | High F1-score across all classes indicates a good balance between precision and recall. |
| Support | 19 | 13 | 13 | --- | 45 | Balanced support across classes indicates a balanced test set. |
| Accuracy | --- | --- | --- | --- | `0.96` | High overall accuracy indicates that the model correctly classified 96% of all instances. |
| Confusion Matrix | 19/19 | `12`/13 | `12`/13 | --- | --- | All instances of class 0 were correctly classified. For classes 1 and 2, one instance each was misclassified. |

**In summary**, the Multinomial Naive Bayes model also performed well on the Iris dataset, with high accuracy and balanced precision and recall for all classes. The model had a slight difficulty with class 1 and class 2, where it misclassified one instance each. The model's high F1-score indicates a good balance between precision and recall. The support shows that the classes are balanced in the test set. The high overall accuracy of 0.96 indicates that the model correctly classified 96% of all instances. The confusion matrix shows that the model had a slight difficulty in distinguishing between classes 1 and 2, leading to one instance of each being misclassified.

### **`Bernoulli Naive Bayes`:**

In [9]:
# model initialize
bnb = BernoulliNB()

# train the model
bnb.fit(X_train, y_train)

# predict the test data
y_pred = bnb.predict(X_test)

# evaluate the model
print("Accuracy Score: ", accuracy_score(y_test, y_pred))
print("Confusion Matrix: \n", confusion_matrix(y_test, y_pred))
print("Classification Report: \n", classification_report(y_test, y_pred))

Accuracy Score:  0.28888888888888886
Confusion Matrix: 
 [[ 0 19  0]
 [ 0 13  0]
 [ 0 13  0]]
Classification Report: 
               precision    recall  f1-score   support

           0       0.00      0.00      0.00        19
           1       0.29      1.00      0.45        13
           2       0.00      0.00      0.00        13

    accuracy                           0.29        45
   macro avg       0.10      0.33      0.15        45
weighted avg       0.08      0.29      0.13        45



c:\Users\Hp\miniconda3\envs\python_ml\Lib\site-packages\sklearn\metrics\_classification.py:1497: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Users\Hp\miniconda3\envs\python_ml\Lib\site-packages\sklearn\metrics\_classification.py:1497: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Users\Hp\miniconda3\envs\python_ml\Lib\site-packages\sklearn\metrics\_classification.py:1497: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


**Observations from the Output**
| Metric | Class 0 | Class 1 | Class 2 | Macro Avg | Weighted Avg | Remarks |
|--------|---------|---------|---------|-----------|--------------|---------|
| Precision | 0.00 | `0.29` | 0.00 | 0.10 | 0.08 | No instances predicted as class 0 and 2 were correct. For class 1, 29% of instances predicted were correct. |
| Recall | 0.00 | `1.00` | 0.00 | 0.33 | 0.29 | No actual instances of class 0 and 2 were identified. All actual instances of class 1 were identified. |
| F1-Score | 0.00 | `0.45` | 0.00 | 0.15 | 0.13 | Low F1-score for class 0 and 2 indicates poor balance between precision and recall. Class 1 has a moderate F1-score. |
| Support | 19 | 13 | 13 | --- | 45 | Balanced support across classes indicates a balanced test set. |
| Accuracy | --- | --- | --- | --- | `0.29` | Low overall accuracy indicates that the model correctly classified only 29% of all instances. |
| Confusion Matrix | `0`/19 | 13/13 | `0`/13 | --- | --- | No instances of class 0 and 2 were correctly classified. All instances of class 1 were correctly classified. |

**In summary**, the Bernoulli Naive Bayes model performed poorly on the Iris dataset, with low accuracy and imbalanced precision and recall for all classes except class 1. The model was unable to correctly classify any instances of class 0 and 2. The model's low F1-score for class 0 and 2 indicates a poor balance between precision and recall, while class 1 has a moderate F1-score. The support shows that the classes are balanced in the test set. The low overall accuracy of 0.29 indicates that the model correctly classified only 29% of all instances. The confusion matrix shows that the model was only able to correctly classify instances of class 1, while it failed to correctly classify any instances of class 0 and 2.

## **`Assumption` should be met for Naive Bayes Algorithm:**
**1. Gaussian Naive Bayes:**
- The features are `continuous`.
- The features are `normally distributed`.
- The features are independent of each other.
- The features have the same variance.
- The features are not correlated with each other.

**2. Multinomial Naive Bayes:**
- The features are `discrete`.
- The features are independent of each other.
- The features are not correlated with each other.
- The features are not negative.
- The features are not zero.
- The features are countable.
- The features are not binary.

**3. Bernoulli Naive Bayes:**
- The features are `binary`.
- The features are independent of each other.
- The features are not correlated with each other.